src/gather.jl

```diff
@@ -49,13 +49,35 @@ function gather_kernel!(dst, src, idx::CUDA.CuDeviceArray{<:CartesianIndex}, max
     return nothing
 end
 
+function checkbounds_src(src, dims::Int, ::Type{<:Any})
+    return i -> checkbounds(Bool, src, ntuple(x -> Colon(), dims)..., i...)
+end
+
+function checkbounds_src(src, dims::Int, ::Type{<:CartesianIndex})
+    return i -> checkbounds(Bool, src, ntuple(x -> Colon(), dims)..., i)
+end
+
 function NNlib.gather!(dst::AnyCuArray, src::AnyCuArray, idx::AnyCuArray)
+    # check dims
     dims = gather_check_dims(src, dst, idx)
     dims_size = size(src)[1:dims]
     max_dims_idx = prod(dims_size)
     max_idx = max_dims_idx * length(idx)
-    args = dst, src, idx, max_idx, max_dims_idx, dims_size
 
+    # check bounds
+    chk = checkbounds_src(src, dims, eltype(idx))
+    in_bnd = map(chk, collect(idx))
```

Comment:
```suggestion
    in_bnd = map(checkbounds_src(src, Val(dims), eltype(idx)), idx)
```
use `Val` should fix the compile error for `ntuple`